tests/unit/test_config_provider.py

```diff
@@ -667,11 +723,102 @@ def _create_os_environ_patcher(self):
 
     def test_config_store_deepcopy(self):
         config_store = ConfigValueStore()
-        config_store.set_config_provider('foo', ConstantProvider('bar'))
+        config_store.set_config_provider(
+            'constant_value', ConstantProvider('ABC')
+        )
         config_store_copy = copy.deepcopy(config_store)
-        config_store_copy.set_config_provider('fizz', ConstantProvider('buzz'))
-        assert config_store.get_config_variable('fizz') is None
-        assert config_store_copy.get_config_variable('foo') == 'bar'
+        config_store_copy.set_config_provider(
+            'constant_value_copy', ConstantProvider('123')
+        )
+        assert config_store.get_config_variable('constant_value_copy') is None
+        assert config_store_copy.get_config_variable('constant_value') == 'ABC'
+
+    # @pytest.fixture
+    # def config_value_store(self):
+    #     return self._create_config_value_store()
+
+    @pytest.mark.parametrize(
+        'defaults_mode',
+        ['standard', 'in-region', 'cross-region', 'mobile', 'auto'],
+    )
+    def test_config_store_providers_not_mutated_after_merge(
+        self, defaults_mode, smart_defaults_factory
+    ):
+        environment_provider = EnvironmentProvider(
+            name='AWS_S3_US_EAST_1_REGIONAL_ENDPOINT',
+            env={'AWS_S3_US_EAST_1_REGIONAL_ENDPOINT': 'regional'},
+        )
+
+        s3_mapping = {
+            'us_east_1_regional_endpoint': ChainProvider(
+                providers=[environment_provider]
+            )
+        }
+
+        config_value_store = self._create_config_value_store(
+            s3_mapping=s3_mapping
+        )
+
+        sts_regional_endpoints_provider = (
+            config_value_store.get_config_provider('sts_regional_endpoints')
+        )
+        s3_section_provider = config_value_store.get_config_provider('s3')
+        retry_mode_provider = config_value_store.get_config_provider(
+            'retry_mode'
+        )
+
+        sts_regional_endpoints_before = (
+            sts_regional_endpoints_provider.provide()
+        )
+        s3_section_before = s3_section_provider.provide()
+        retry_mode_before = retry_mode_provider.provide()
+
+        smart_defaults_factory.merge_smart_defaults(
+            config_value_store, defaults_mode, 'some-region'
+        )
+
+        sts_regional_endpoints_after = config_value_store.get_config_variable(
+            'sts_regional_endpoints'
+        )
+        retry_mode_after = config_value_store.get_config_variable('retry_mode')
+
+        assert (
+            sts_regional_endpoints_provider.provide()
+            == sts_regional_endpoints_before
```

Comment:
And when we parameterize the config variable names, let's also add explicit parameter values for what we expect the before and after values are expected to be. When we generalize it like this it make it easy to cover up issues in the test where the values can match our expectations (e.g. the before and after values are not equal) but the values themselves are not as expected (e.g. one of the values is `None` instead of a string value).